Add TSPLIB95-importer

HannesWell opened the pull request:
I created an importer to read symmetric `travelling salesman problem` instances from files in the `TSPLIB95` format.
Files in TSPLIB95 format contain exactly defined TSP instances. As stated in the Javadoc of the importer there are large TSPLIB95 libraries containing problem instances of different magnitudes provided by the University of [Heidelberg](http://comopt.ifi.uni-heidelberg.de/software/TSPLIB95/) or [Waterloo](http://www.math.uwaterloo.ca/tsp/data/index.html).

The instances provided there can be used as unified benchmarks or tests for TSP algorithms. 
Additionally the importer could be used to improve the tests of JGraphT's TSP algorithms. At the moment some tests use randomly created instances that are therefore different for each test run. Such instances could be stored in a TSPLIB95 file and then load each time the test is executed to have the same instance every time. Also TSP test instances with special properties could be defined to test certain aspects or corner cases of an algorithm.

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+ */
+public class TSPLIBImporter<V>
+    implements
+    GraphImporter<V, DefaultWeightedEdge>
```
Unless absolutely necessary, we try to not fix the edge class for the user. Change this to a generic parameter `E`. This also affects your parallel streams loading later on in your code.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+     * @author Hannes Wellmann
+     * @param <V> the type of vertices in the imported graph
+     */
+    public static class TSPLIBFileData<V>
```
I would prefer a nicer name here. You are already inside class `TSPLIBImporter` which means that you do not have to prefix it with `TSPLIB` anymore. Additionally, since this information is data about data, I would use something like  a `Metadata` or something in this line.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+        private boolean hasDistinctLocations;
+        private Boolean hasDistinctEdges;
+
+        private Graph<V, DefaultWeightedEdge> graph;
```
If you are only using this for `hasDistinctEdgesPerVertex` then ok, otherwise this is not needed.
The user provides the graph from the `importGraph` method, no need to include it back.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+         * 
+         * @return the complete {@code Graph} of location connections
+         */
+        public Graph<V, DefaultWeightedEdge> getGraph()
```
Remove this method.  We have a clear interface on all importers, which is to the `importGraph` method. No need to return the graph using two different ways.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+         * 
+         * @return the tour described by the list of vertex indices
+         */
+        public List<Integer> getTour()
```
What does it mean to return a list of integers `List<Integer>`?  Vertices in JGraphT may be any type of object, you need to translate the integers to vertices.

HannesWell replied:
The problem is that a TSPLIB-file containing a tour does not need to contain the node coordinates defining the TSP-problem instance.
In the cases I know there is a file containing the `NODE_COORD_SECTION` with the coordinates of all nodes (where each node also has a number) and a separate file that contains for example the optimal tour or the best known tour. That tour is described by a List of node-numbers that correspond to the numbers in the file with the `NODE_COORD_SECTION`. It is possible but definitely not always the case that the node coordinates and the tour is contained in one file.

I agree that a List of vertices is better. I thought about multiple options to achieve this. The one I prefer is the following:

- Let `getTour()` return a `List<V>` that is only not null, if a file contains a `TOUR_SECTION`.
- create a separate method `importTour(List<V>)` that has a List of vertices as their parameter that is ordered according to the node numbers and returns the List of vertices in the order of the tour. This enables import from a file that only contains a `TOUR_SECTION` . I think this method should create a MetaData object too (but maybe this has to be discussed).
- To enable the latter method I would also add a `getOrderedVertices()` method to the MetaData class that returns the vertices imported from a file with `NODE_COORD_SECTION` in their "number"-order.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+     * Read the complete {@link Graph} of node locations and returns true if all locations are
+     * distinct.
+     */
+    private boolean readCompleteNodeGraphData(
```
Please merge methods `readCompleteNodeGraphData`, `readNodeCoordinateData` and `buildCompleteGraph` into a single method called 'readGraph` or something like this.

I know that best practices are to keep methods small, but in this case it is over complicating things.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+        // Enable parallel streams. Synchronization not necessary, because DefaultWeightedEdge are
+        // IntrusiveEdges and modified directly.
+
+        graph.edgeSet().parallelStream().forEach(e -> {
```
Remove the parallel stream here, in order to also avoid any problems with using a generic edge type `E`. 

In the last few years, we explicitly try to avoid code which makes assumptions about the user edge.

d-michail replied:
Moreover, these are probably small instances since the TSP is a hard problem, giving no real motivation for parallel processing.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+        double getElement(V vector, int index);
+    }
+
+    private final Function<double[], V> vectorFactory;
```
Completely remove the `vectorFactory` and the `elementGetter` abstractions. 

Graphs in JGraphT are build using the corresponding `vertexSupplier` inside the graph instance.

After you create the vertices using method `addVertex()`, you need to internally maintain a map from the vertices to the coordinates (read from file) of the vertex. Afterwards you need to iterate over all pair of vertices, acquire from the map the coordinates pair and call the corresponding distance method.

HannesWell replied:
All right. I agree with your remarks and will apply them in the next days.

The intention for the factory was to be able to set the coordinates read from the file to the vertices. I assumed that this importer would likely be used with some kind of vector data structure. But it looks like I was too specific here and I understand your point.

Nevertheless I would like to provide a user the possibility to have the read coordinates in the created vertices. What do you think about a `BiConsumer` to set the read coordinates into a vertex that a user can optionally supply in the constructor? The vertex would be created by the regular vertex supplier and the read coordinates would be stored Map as you suggested. Afterwards (if the setter is present) for each vertex its coordinates could be set to it with the setter. The subsequent edge computations would fully operate on the internal coordinate data structure of the importer.

HannesWell replied:
Or alternatively I provide the unmodifiable Vertex-Coordinates-Map in the MetaData-object. I think is better because simpler.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+     * @param vectorFactory used to create vector vertices with certain element values
+     * @param elementGetter used to obtain element values from the created vertex-objects
+     */
+    public TSPLIBImporter(Function<double[], V> vectorFactory, ElementGetter<V> elementGetter)
```
No need to any parameter here.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+
+    // import data
+
+    private TSPLIBFileData<V> lastImportData = null;
```
Remove the `last` and the `=null` since Java does this automatically.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+     * 
+     * @return {@code TSPLIBFileData} of the latest import
+     */
+    public TSPLIBFileData<V> getLastImportData()
```
Call this like `getMetadata()` or something similar.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+     * @throws IllegalArgumentException if the specified {@code graph} is not simple or is not empty
+     */
+    @Override
+    public void importGraph(Graph<V, DefaultWeightedEdge> graph, Reader in)
```
DefaultWeightedEdge -> E

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+        lastImportData = null;
+        vectorLength = -1;
+        if (graph == null) {
+            @SuppressWarnings("unchecked") Supplier<DefaultWeightedEdge> edgeSupplier =
```
The graph is not supposed to be null here. It is not your responsibility to create a graph here.

Now I understand why you return the graph inside the metadata object. Please remove this pattern. We want to keep all importers working in a similar way.

d-michail commented on jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java:
```diff
@@ -0,0 +1,735 @@
+            if ("-1".equals(lineContent)) {
+                break;
+            }
+            tour.add(Integer.valueOf(lineContent));
```
You need to consult some `Map<Integer, V>` here which you probably need to build earlier in your code in order to create a tour which is a `List<V>`.

HannesWell commented:
I revised the structure of the importer and adjusted the tests accordingly.
I agree with you, the importer was very specific and I tried to make it more generic and addressed the other points you mentioned. I think the new structure is suitable to fit the TSPLIB-format into a `GraphImporter` in a generic way.

HannesWell commented:
Another question: Have you ever considered to include Hamcrest as dependency for tests? Their matchers are quiet handy in combination with JUnit. Or do you want to keep the number of dependencies small?

jsichi commented:
Test dependencies are less of a concern, so I wouldn't be opposed to adding hamcrest matchers.

d-michail commented:
I will take a look in the next few days. Sorry for the delay.

HannesWell commented:
No problem. I'm also quite busy those days.

d-michail commented:
Great! The final code ended up looking pretty nice!

HannesWell commented:
Thank you. I'm happy to hear that.

HannesWell commented:
Regarding the Hamcrest Matchers:
Do you want the Maven dependency to Hamcrest to be add in a separate PR or as a "side-effect" in a PR where they are used? In the first case the question would be whether it should be add to all jGraphT artefacts (and probably would not be used in the first place)?

d-michail commented:
I guess better in a PR where they are also used.
